nomad/heartbeat.go

```diff
@@ -172,21 +174,44 @@ func (h *nodeHeartbeater) invalidateHeartbeat(id string) {
 	}
 }
 
-func (h *nodeHeartbeater) shouldDisconnect(id string) bool {
+func (h *nodeHeartbeater) disconnectState(id string) (bool, bool) {
+	node, err := h.State().NodeByID(nil, id)
+	if err != nil {
+		h.logger.Error("error retrieving node by id", "error", err)
+		return false, false
+	}
+
+	// Exit if this is the node already in a state other than ready.
+	if node.Status != structs.NodeStatusReady {
+		return false, false
+	}
+
 	allocs, err := h.State().AllocsByNode(nil, id)
 	if err != nil {
 		h.logger.Error("error retrieving allocs by node", "error", err)
-		return false
+		return false, false
 	}
 
 	now := time.Now().UTC()
+	// Check if the node has any allocs that are configured with max_client_disconnect,
+	// and if so, whether it has pending reconnects. If one alloc isn't yet expired
+	// it still has pending reconnects.
```

Comment:
I think this is more accurate in terms of the first conditional below?

```suggestion
	// Check if the node has any allocs that are configured with max_client_disconnect,
	// that are past the disconnect window, and if so, whether it has at least one alloc
	// that isn't yet expired.
```